### ISSUE TITLE
StackOverflowError in parser with Settings.parseWithMetaLookups == IGNORE_ON_FAILURE

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11856]
This test helps reproduce a `StackOverflowError` in 3.15.0 and 3.14.9:

```java
@Test
public void x() {
    String sql = """
        select b
        from (
          select x a,
            y b
          from (
            select
              a x,
              b y
            from t1
          ) t2
        ) t3
        """;

    dsl.settings().setParseWithMetaLookups(ParseWithMetaLookups.IGNORE_ON_FAILURE);
    Meta meta = dsl.meta("");
    dsl.configuration().set((MetaProvider) () -> meta);
    System.out.println(dsl.parser().parseSelect(sql));
}
```

The problem seems to be related to renaming columns from `a -> x -> a` in derived tables. The relevant part of the stack being:

```
DataTypeProxy<T>.getBinding() line: 128	
DataTypeProxy<T>.getBinding() line: 128	
DataTypeProxy<T>.getBinding() line: 128	
DataTypeProxy<T>.getBinding() line: 128	
DataTypeProxy<T>.getBinding() line: 128	
FieldAlias<T>(AbstractTypedNamed<T>).getBinding() line: 73	
TableAlias<R>.lambda$2(Name[], Field, int) line: 89	
0x0000000800dcbc58.apply(Object, int) line: not available	
Tools.map(T[], ThrowingObjIntFunction<? super T,? extends U,E>) line: 2010	
TableAlias<R>.init(Name[]) line: 86	
TableAlias<R>.<init>(Table<R>, Name, Name[], Predicate<Context<?>>) line: 78	
TableAlias<R>.<init>(Table<R>, Name, Predicate<Context<?>>) line: 67	
DerivedTable<R>.as(Name) line: 71	
DerivedTable<R>(AbstractTable<R>).as(String) line: 405	
SelectQueryImpl<R>.asTable(String) line: 824	
DefaultParseContext.t(TableLike<R>, boolean) line: 6075	
DefaultParseContext.parseTableFactor() line: 6351	
DefaultParseContext.parseLateral() line: 6063	
DefaultParseContext.parseTable() line: 6048	
0x0000000800ddefc0.apply(Object) line: not available	
DefaultParseContext.parseList(Predicate<ParseContext>, Function<ParseContext,? extends T>) line: 12153	
DefaultParseContext.parseList(char, Function<ParseContext,? extends T>) line: 12140	
DefaultParseContext.parseQueryPrimary(Integer, WithImpl) line: 1337	
DefaultParseContext.parseQueryTerm(Integer, WithImpl, SelectQueryImpl<Record>) line: 1195	
DefaultParseContext.parseQueryExpressionBody(Integer, WithImpl, SelectQueryImpl<Record>) line: 1161	
DefaultParseContext.parseSelect(Integer, WithImpl) line: 937	
DefaultParseContext.parseSelect() line: 932	
DefaultParseContext.parseQuery(boolean, boolean) line: 778	
DefaultParseContext.parseSelect0() line: 565	
ParserImpl.parseSelect(String, Object...) line: 418	
ParserImpl.parseSelect(String) line: 413	
ParserImplTest.x() line: 121	
```

**Comment by lukaseder** [comment]
There's a cycle in `DataTypeProxy` references:

![image](https://user-images.githubusercontent.com/734593/117325891-7157c300-ae91-11eb-9584-9e7322327e5d.png)

**Comment by lukaseder** [comment]
Weirdly, it doesn't happen with only a single column:

```java
@Test
public void testParseWithMetaLookupsCircularAliasing() {
    String sql = """
        select b
        from (
          select
            y b
          from (
            select
              b y
            from t1
          ) t2
        ) t3
        """;

    dsl.settings().setParseWithMetaLookups(ParseWithMetaLookups.IGNORE_ON_FAILURE);
    Meta meta = dsl.meta("");
    dsl.configuration().set((MetaProvider) () -> meta);
    System.out.println(dsl.parser().parseSelect(sql));
}
```

**Comment by lukaseder** [comment]
See also: https://github.com/jOOQ/jOOQ/issues/14343

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.6 (https://github.com/jOOQ/jOOQ/issues/14344), and 3.16.12  (https://github.com/jOOQ/jOOQ/issues/14345)

**Comment by lukaseder** [comment]
This error and some related ones (e.g. https://github.com/jOOQ/jOOQ/issues/14373), are only caused by the fact that we currently run a single pass through the parser to parse SQL and resolve identifiers. This is a bit brittle. It was a conscious decision to use this algorithm because the `QOM` API wasn't ready yet, when we started supporting meta lookups.

Once the `QOM` API is ready, we'll refactor this logic in the parser, and resolve identifiers only once everything has been parsed, in a second pass. That will also remove all the mutability in `FieldProxy` and `DataTypeProxy`, which will then no longer be necessary.
